### PR TITLE
Fixes two crew objectives and speeds up checks for several others.

### PR DESCRIPTION
**Pull request by 112358sam**
[bug]

## About the PR 
Fixes the "get the head surgeon to centcom" and "leave all station furnaces on" crew objectives. The former now correctly checks for either the special cardboard box or special medbot from an appropriate list rather than checking for the cardboard box from the list of bots. The latter will now fail if any furnace on station is off rather than succeeding if at least one furnace is on.

In addition, all `for(<type> in world)` loops have been replaced with `by_type`'s or `by_cat`'s to reduce server load.

## Why's this needed? 
Two of the crew objectives were not working as stated. This fixes that. In addition, looping through every object in game should be avoided. This will hopefully reduce the end-of-round lag when all the crew objectives get checked.

### CODE REVIEW

**Comment by ZeWaka** [comment]
are these actually faster? `in world` has some hidden typecaching

**Comment by 112358sam** [comment]
> 
> 
> are these actually faster? I'd like to see some numbers/stats. `in world` has some hidden typecaching

I had no idea, we had a discussion on imcoder and we thought `in world` iterated through pretty much everything. Do you know where I could find more information on how `in world` works?

**Comment by 112358sam** [comment]
Here is the performance before and after the change. Forgive the poor formatting. I just screenshot the values and threw them together in one image.
image

**Comment by pali6** [comment]
> are these actually faster? I'd like to see some numbers/stats. `in world` has some hidden typecaching

I believe that only works on the level of direct /atom subtypes and not anything deeper.

**Comment by ZeWaka** [comment]
> > are these actually faster? I'd like to see some numbers/stats. `in world` has some hidden typecaching
> 
> I believe that only works on the level of direct /atom subtypes and not anything deeper.

yeah that's what i was thinking of - alas

File: code/modules/hydroponics/plants_herb.dm [hunk 1]
```diff
@@ -136,6 +136,14 @@ ABSTRACT_TYPE(/datum/plant/herb)
 	/datum/plantmutation/cannabis/white,/datum/plantmutation/cannabis/ultimate)
 	commuts = list(/datum/plant_gene_strain/resistance_drought,/datum/plant_gene_strain/yield/stunted)
 
+	New()
+		. = ..()
+		START_TRACKING_CAT(TR_CAT_CANNABIS_OBJ_ITEMS)
+
+	disposing()
+		. = ..()
```
Comment (pali6) [inline code/modules/hydroponics/plants_herb.dm:144]: Generally in disposing I'd prefer calling parent to be the last action. Think if disposing as the closing bracket to New being the opening bracket off the object. It doesn't make a change here but it's good practice.

Comment (pali6) [reply code/modules/hydroponics/plants_herb.dm:144]: Applies to other disposing in this PR too.

**Review by pali6 — COMMENTED** [review]
Good changes

File: code/datums/crew_objective.dm [hunk 2]
```diff
@@ -147,9 +147,9 @@ ABSTRACT_TYPE(/datum/objective/crew/chiefengineer)
 		medal_name = "Slow Burn"
 		check_completion()
 			for(var/obj/machinery/power/furnace/F in machine_registry[MACHINES_POWER])
-				if(F.z == 1 && F.active == 1)
-					return 1
-			return 0
+				if(F.z == 1 && F.active == 0 && istype(F.loc.loc, /area/station))
```
Comment (Zonespace27) [inline code/datums/crew_objective.dm:150]:
```suggestion
				if(F.z == 1 && !F.active && istype(F.loc.loc, /area/station))
```

File: code/modules/hydroponics/plants_herb.dm (hunk 1 above)
Comment (112358sam) [reply code/modules/hydroponics/plants_herb.dm]: Done!

**Review by pali6 — APPROVED** [review]

**Pull request merged**